src/com/ms/silverking/cloud/dht/management/aws/MultiInstanceLauncher.java

```diff
@@ -36,284 +36,292 @@
-        
-        printDone( getInstanceIds(workerInstances) );
+  private Instance launchInstance;
+  private String launchKeyPairName;
+  private List<GroupIdentifier> securityGroups;
+  private String subnetId;
+
+  public static final String privateKeyFilename = USER_HOME + "/.ssh/id_rsa";
+  public static final String ipsFilename = SKCloudAdmin.cloudOutDir + "/cloud_ip_list.txt";
+  private String privateKey;
+
+  private static final String newSecurityGroupName = "sk_instance";
+
+  private List<Instance> workerInstances;
+
+  public MultiInstanceLauncher(AmazonEC2 ec2, String launchHostIp, int numInstances, String amiId, String instanceType,
+      boolean includeMaster) {
+    this.ec2 = ec2;
+    this.launchHostIp = launchHostIp;
+    this.keyPairName = getUniqueKeyPairName(launchHostIp);
+    this.amiId = amiId;
+    this.instanceType = instanceType;
+    this.includeMaster = includeMaster;
+
+    // needs to be after includeMaster
+    Util.checkNumInstances(numInstances);
+    setNumWorkerInstances(numInstances);
+
+    workerInstances = Collections.emptyList();
+  }
+
+  private void setNumWorkerInstances(int numInstances) {
+    numWorkerInstances = numInstances;
+    if (includeMaster)
+      numWorkerInstances--;
+  }
+
+  public String getKeyPairName() {
+    return keyPairName;
+  }
+
+  public void run() {
+    System.out.println("Creating " + numWorkerInstances + " new instance(s)");
+
+    checkIamRoleIsAttached();
+
+    if (!isMasterOnlyInstance()) {
+      setLaunchInstance();
+      //        createSecurityGroup();
+      //        addSecurityGroupToLaunchInstance();
     }
-    
-    private List<String> getSecurityGroupIds(List<GroupIdentifier> securityGroups) {
-        List<String> names = new ArrayList<>();
-        
-        for (GroupIdentifier group : securityGroups)
-            names.add(group.getGroupId());
-        
-        return names;
+
+    deleteKeyPair(ec2, keyPairName);
+    createKeyPair();
+    createPrivateKeyFile();
+
+    if (!isMasterOnlyInstance()) {
+      createAndRunNewInstances();
+      waitForInstancesToBeRunning(ec2, workerInstances);
+      waitForInstancesToBeReachable(ec2, workerInstances);
     }
-    
-//    private List<String> getNames(List<GroupIdentifier> securityGroups) {
-//        List<String> names = new ArrayList<>();
-//        
-//        for (GroupIdentifier group : securityGroups)
-//            names.add(group.getGroupName());
-//        
-//        return names;
-//    }
-    
-    private void createIpListFile() {
-        print("Creating IpList File");
-        
-        writeToFile(ipsFilename, getInstanceIps());
-        
-        printDone(ipsFilename);
+
+    createIpListFile();
+  }
+
+  public void checkIamRoleIsAttached() {
+    DescribeInstancesRequest request = new DescribeInstancesRequest();
+    try {
+      DescribeInstancesResult response = ec2.describeInstances(request);
+    } catch (SdkClientException exception) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)